File: bindings/imgui_bundle/demos_python/demos_imguizmo/demo_guizmo_gradient.py
# Demo gradient with ImGuizmo
# See equivalent C++ program: demos_cpp/demos_imguizmo/demo_guizmo_gradient.main.cpp
# Hum... I don't get the point of this widget

from typing import List
from imgui_bundle.demos_python.demo_utils.api_demos import GuiFunction
from imgui_bundle import imgui, ImVec4, ImVec2, imguizmo, immapp


def add_imvec4(a: ImVec4, b: ImVec4) -> ImVec4:
    return ImVec4(a.x + b.x, a.y + b.y, a.z + b.z, a.w + b.w)


def mul_scalar_imvec4(a: ImVec4, k: float) -> ImVec4:
    return ImVec4(a.x * k, a.y * k, a.z * k, a.w * k)


ImVec4.__add__ = add_imvec4  # type: ignore # monkey patching
ImVec4.__mul__ = mul_scalar_imvec4  # type: ignore # monkey patching


class MyGradient(imguizmo.im_gradient.DelegateStl):  # type: ignore
    points: List[ImVec4]

    def __init__(self) -> None:
        imguizmo.im_gradient.DelegateStl.__init__(self)
        # The last value (ImVec4.w) stores the position on a line
        nb_elems = 4.0
        pos = 0.0
        dpos = 1.0 / (nb_elems - 1.0)
        self.points = []
        self.points.append(ImVec4(1, 1, 1, pos))
        pos += dpos
        self.points.append(ImVec4(0, 1, 1, pos))
        pos += dpos
        self.points.append(ImVec4(1, 0, 1, pos))
        pos += dpos
        self.points.append(ImVec4(1, 1, 0, pos))
        pos += dpos

    def get_points_list(self) -> List[ImVec4]:
        return self.points

    def edit_point(
        self, point_index: int, value: ImVec4
    ) -> int:  # overridable (pure virtual)
        self.points[point_index] = value
        return 0

    def get_point(self, t: float) -> ImVec4:
        sorted_values = self.sorted_values()
        if t <= 0.0:
            return sorted_values[0]
        elif t >= 1.0:
            return sorted_values[-1]

        idx = len(sorted_values) - 1
        while (idx >= 1) and (t < sorted_values[idx].w):
            idx -= 1

        v0 = sorted_values[idx]
        v1 = sorted_values[idx + 1]
        interval_length = v1.w - v0.w
        k0 = (t - v0.w) / interval_length
        k1 = 1.0 - k0
        r = v0 * k0 + v1 * k1  # type: ignore
        return r  # type: ignore

    def add_point(self, value: ImVec4) -> None:
        self.points.append(value)

    def sorted_values(self) -> List[ImVec4]:
        r = sorted(self.points, key=lambda p: p.w)
        return r


# This returns a closure function that will later be invoked to run the app
def make_closure_demo_guizmo_gradient() -> GuiFunction:
    my_gradient = MyGradient()
    size = ImVec2(400, 20)

    def gui() -> None:
        _result = imguizmo.im_gradient.edit_pure(my_gradient, size)
        imgui.text_wrapped(
            """
            I'm not sure about the purpose of this widget.
            You can drag squares, and double click to add some more
        """
        )

    return gui


def main():
    gui = make_closure_demo_guizmo_gradient()

    # Run app
    immapp.run(gui, window_size=(400, 100))


if __name__ == "__main__":
    main()

File: bindings/imgui_bundle/demos_cpp/demos_imguizmo/demo_guizmo_gradient.cpp
#ifdef IMGUI_BUNDLE_WITH_IMGUIZMO
// Demo gradient with ImGuizmo
// See equivalent python program: bindings/imgui_bundle/demos/demos_imguizmo/Wrapper/demo_guizmo_gradient.py
// Hum... I don't get the point of this widget

#include "demo_utils/api_demos.h"

#define IMGUI_DEFINE_MATH_OPERATORS
#include "imgui.h"
#include "immapp/immapp.h"
#include "ImGuizmoPure/ImGradientPure.h"

#include <algorithm>


static ImVec4 operator*(const ImVec4& lhs, float rhs)
{ return ImVec4(lhs.x * rhs, lhs.y * rhs, lhs.z * rhs, lhs.w * rhs); }


struct MyGradient: public ImGradient::DelegateStl
{
    MyGradient()
    {
        // The last value (ImVec4.w) stores the position on a line
        int nb_elems = 4.;
        float pos = 0.f;
        float dpos = 1.f / (nb_elems - 1.f);

        mPoints.push_back(ImVec4(1,1,1,pos)); pos += dpos;
        mPoints.push_back(ImVec4(0,1,1,pos)); pos += dpos;
        mPoints.push_back(ImVec4(1,0,1,pos)); pos += dpos;
        mPoints.push_back(ImVec4(1,1,0,pos)); pos += dpos;
    }
    std::vector<ImVec4>& GetPointsList() override
    {
        return mPoints;
    }
    int EditPoint(int pointIndex, ImVec4 value) override
    {
        mPoints[pointIndex] = value;
        return 0;
    }
    ImVec4 GetPoint(float t) override
    {
        auto sortedValues = SortedValues();
        if (t <= 0.)
            return sortedValues.front();
        else if (t >= 1.)
            return sortedValues.back();

        size_t idx = sortedValues.size() - 1;
        while ((idx >= 1) && (t < sortedValues[idx].w))
            --idx;

        const auto& v0 = sortedValues[idx];
        const auto& v1 = sortedValues[idx + 1];
        float interval_length = v1.w - v0.w;
        float k0 = (t - v0.w) / interval_length;
        float k1 = 1.f - k0;
        ImVec4 r = v0 * k0 + v1 * k1;
        return r;
    }
    void AddPoint(ImVec4 value) override
    {
        mPoints.push_back(value);
    }

    std::vector<ImVec4> SortedValues()
    {
        std::vector<ImVec4> r = mPoints;
        std::sort(r.begin(), r.end(), [](ImVec4 a, ImVec4 b) { return a.w < b.w; });
        return r;
    }

    std::vector<ImVec4> mPoints;
};


 void demo_guizmo_gradient()
{
    static MyGradient myGradient;
    ImVec2 size(400.f, 20.f);
    ImGradient::EditPure(myGradient, size);
    ImGui::TextWrapped(R"(
        I'm not sure about the purpose of this widget.
        You can drag squares, and double click to add some more)");
}

#else // #ifdef IMGUI_BUNDLE_WITH_IMGUIZMO
#include "imgui.h"
void demo_guizmo_gradient() { ImGui::Text("This demo requires ImGuizmo."); }
#endif // IMGUI_BUNDLE_WITH_IMGUIZMO